sdc11073/wsdiscovery.py

```diff
@@ -813,126 +828,132 @@ def removeSourceAddr(self, addr):
+                self._logger.warning('socket read error %s', e)
                 time.sleep(0.01)
                 continue
             if self.isFromMySocket(addr):
                 continue
-            getCommunicationLogger().logDiscoveryMsgIn(addr[0], data)
+            self._read_queue.put((addr, data))
 
-            env = None
-            env = parseEnvelope(data, addr[0])
-            if env is None:  # fault or failed to parse
-                continue
 
-            mid = env.getMessageId()
-            if mid in self._knownMessageIds:
-                self._logger.debug('message Id %s already known. This is a duplicate receive, ignoring.', mid)
-                continue
+    def _run_q_read(self):
+        """Read from internal queue and process message"""
+        while not self._quitRecvEvent.is_set():
+            try:
+                incoming = self._read_queue.get(timeout=0.1)
+            except queue.Empty:
+                pass
             else:
-                self._knownMessageIds.appendleft(mid)
-                try:
-                    del self._knownMessageIds[-50]  # limit length of remembered message Ids
-                except IndexError:
-                    pass
-            iid = env.getInstanceId()
-            mid = env.getMessageId()
-            if iid:
-                mnum = env.getMessageNumber()
-                key = addr[0] + ":" + str(addr[1]) + ":" + str(iid)
-                if mid is not None and len(mid) > 0:
-                    key = key + ":" + mid
-                if not key in self._iidMap:
-                    self._iidMap[key] = iid
+                addr, data = incoming
+                getCommunicationLogger().logDiscoveryMsgIn(addr[0], data)
+
+                env = parseEnvelope(data, addr[0], self._logger)
+                if env is None:  # fault or failed to parse
+                    continue
+
+                mid = env.getMessageId()
+                if mid in self._knownMessageIds:
+                    self._logger.debug('message Id %s already known. This is a duplicate receive, ignoring.', mid)
+                    continue
                 else:
-                    tmnum = self._iidMap[key]
-                    if mnum > tmnum:
-                        self._iidMap[key] = mnum
+                    self._knownMessageIds.appendleft(mid)
+                    try:
+                        del self._knownMessageIds[-50]  # limit length of remembered message Ids
+                    except IndexError:
+                        pass
+                iid = env.getInstanceId()
+                mid = env.getMessageId()
+                if iid:
+                    mnum = env.getMessageNumber()
+                    key = addr[0] + ":" + str(addr[1]) + ":" + str(iid)
+                    if mid is not None and len(mid) > 0:
+                        key = key + ":" + mid
+                    if not key in self._iidMap:
+                        self._iidMap[key] = iid

```

Comment:
I think this map is a memory leak, if instance id's are used, it won't ever be cleaned up. Is this intentionally? The message id's are wiped at some point.